File: api/v1/endpoints/gold.py
# -*- coding: utf-8 -*-
"""
===================================
API Vàng (Phase A — đa tài sản)
===================================

GET /api/v1/gold/overview — giá vàng SJC, giá vàng thế giới, tỷ giá USD/VND
và chênh lệch (premium) SJC so với thế giới quy đổi sang VND/lượng.

Logic lấy dữ liệu + tính toán + cache nằm ở src/services/gold_service.py.
"""

import logging

from fastapi import APIRouter, Query

from api.v1.schemas.gold import GoldHistoryResponse, GoldOverviewResponse

logger = logging.getLogger(__name__)

router = APIRouter()


@router.get("/overview", response_model=GoldOverviewResponse)
def gold_overview() -> GoldOverviewResponse:
    """Tổng quan vàng + bảng loại vàng (cache 5 phút trong tiến trình)."""
    from src.services.gold_service import get_gold_overview

    return GoldOverviewResponse(**get_gold_overview())


@router.get("/history", response_model=GoldHistoryResponse)
def gold_history(
    days: int = Query(180, ge=30, le=365, description="Số ngày lịch sử"),
    step_days: int = Query(14, ge=7, le=30, description="Khoảng cách giữa các mốc"),
) -> GoldHistoryResponse:
    """Lịch sử giá vàng SJC + thế giới quy đổi (VND/lượng). Cache 6 giờ.

    Lần gọi nguội lấy dữ liệu nhiều mốc nên có thể mất vài giây.
    """
    from src.services.gold_service import get_gold_history

    return GoldHistoryResponse(**get_gold_history(days=days, step_days=step_days))

File: web/src/api/gold.ts
import apiClient from './index'
import { toCamelCase } from './utils'
import type { GoldOverview, GoldHistory } from '@/types/gold'

export const goldApi = {
  /** Tổng quan vàng: giá SJC, giá thế giới, tỷ giá, chênh lệch (premium), bảng loại vàng. */
  async getOverview(): Promise<GoldOverview> {
    const res = await apiClient.get<Record<string, unknown>>('/api/v1/gold/overview')
    return toCamelCase<GoldOverview>(res.data)
  },

  /** Lịch sử giá vàng SJC + thế giới quy đổi (VND/lượng). Lần đầu có thể chậm vài giây. */
  async getHistory(days = 180, stepDays = 14): Promise<GoldHistory> {
    const res = await apiClient.get<Record<string, unknown>>('/api/v1/gold/history', {
      params: { days, step_days: stepDays },
    })
    return toCamelCase<GoldHistory>(res.data)
  },
}
